In [1]:
import math
import numpy as np
from scipy.stats import t

## 📌 Question 1 : One-Sample t-Test (Mean)

A sample of 20 students scored: mean = 72, s = 8. The national average is 70. At α = 0.05, is this school's performance significantly different? Use a one-sample t-test. Compute t, df, the critical value, and state your conclusion.

#### Step 1 : Hypotheses

- **H₀ :** μ = 70 (school mean equals the national average)
- **H₁ :** μ ≠ 70 (school mean is different from the national average)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

Population standard deviation (σ) is unknown and the sample is small (n = 20), so the t-distribution is used instead of Z. With one sample being compared to a known population mean, this is a one-sample t-test.

#### Step 3 : Test Statistic

$$t = \frac{\bar{x} - \mu_0}{s / \sqrt{n}}$$

Degrees of freedom: df = n - 1 = 19

In [2]:
n1 = 20          # sample size
xbar1 = 72       # sample mean
s1 = 8           # sample standard deviation
mu1 = 70         # null hypothesis population mean
alpha1 = 0.05    # significance level
df1 = n1 - 1     # degrees of freedom

In [3]:
t1 = (xbar1 - mu1) / (s1 / np.sqrt(n1))  # one-sample t-statistic
t1

np.float64(1.118033988749895)

#### Step 4 : Critical Value Method

In [4]:
tcrit1 = t.ppf(1 - alpha1 / 2, df1)  # two-tailed critical value at alpha=0.05, df=19
tcrit1

np.float64(2.093024054408263)

**|t| = 1.118 < t_crit = 2.093 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [5]:
pval1 = 2 * t.sf(abs(t1), df1)  # two-tailed p-value
pval1

np.float64(0.2774901395757606)

**p-value = 0.2775 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [6]:
se1 = s1 / np.sqrt(n1)           # standard error
lo1 = xbar1 - tcrit1 * se1       # lower bound of 95% CI
hi1 = xbar1 + tcrit1 * se1       # upper bound of 95% CI
print(f"95% CI: ({lo1:.4f}, {hi1:.4f})")

95% CI: (68.2559, 75.7441)


### ✅ Conclusion

| Metric | Value |
|---|---|
| Sample mean | 72 |
| Null mean (μ₀) | 70 |
| t-statistic | 1.1180 |
| df | 19 |
| Critical value (±) | 2.0930 |
| p-value | 0.2775 |
| 95% CI | (68.2559, 75.7441) |
| Decision | Fail to reject H₀ |

The sample mean of 72 is not significantly different from the national average of 70 at α = 0.05. The t-statistic of 1.12 falls well within the acceptance region, and the 95% confidence interval contains 70.

---

## 📌 Question 2 : Two-Sample t-Test (Pooled, Equal Variances)

Drug Group (n = 15, x̄ = 5.2, s = 1.1) vs Placebo Group (n = 15, x̄ = 4.8, s = 0.9). Test whether the drug has a significant effect at α = 0.05. Assume equal variances. Compute the pooled standard deviation and the t-statistic manually.

#### Step 1 : Hypotheses

- **H₀ :** μ₁ = μ₂ (drug and placebo groups have the same mean)
- **H₁ :** μ₁ ≠ μ₂ (drug group mean differs from placebo)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

Two independent groups are being compared, population standard deviations are unknown, and equal variances are assumed. This calls for a two-sample pooled t-test. Pooling the variances gives a more precise estimate of the common population variance when the equal-variance assumption holds.

#### Step 3 : Test Statistic

Pooled standard deviation:

$$s_p = \sqrt{\frac{(n_1 - 1)s_1^2 + (n_2 - 1)s_2^2}{n_1 + n_2 - 2}}$$

t-statistic:

$$t = \frac{\bar{x}_1 - \bar{x}_2}{s_p \sqrt{\frac{1}{n_1} + \frac{1}{n_2}}}$$

Degrees of freedom: df = n₁ + n₂ - 2 = 28

In [7]:
na2 = 15          # drug group sample size
xbar2a = 5.2      # drug group sample mean
sa2 = 1.1         # drug group sample standard deviation
nb2 = 15          # placebo group sample size
xbar2b = 4.8      # placebo group sample mean
sb2 = 0.9         # placebo group sample standard deviation
alpha2 = 0.05     # significance level
df2 = na2 + nb2 - 2  # degrees of freedom for pooled two-sample t-test

In [8]:
sp2 = math.sqrt(((na2 - 1) * sa2**2 + (nb2 - 1) * sb2**2) / (na2 + nb2 - 2))  # pooled standard deviation
sp2

1.004987562112089

In [9]:
se2 = sp2 * np.sqrt(1 / na2 + 1 / nb2)   # pooled standard error
t2 = (xbar2a - xbar2b) / se2              # two-sample pooled t-statistic
t2

np.float64(1.0900086292691404)

#### Step 4 : Critical Value Method

In [10]:
tcrit2 = t.ppf(1 - alpha2 / 2, df2)  # two-tailed critical value at alpha=0.05, df=28
tcrit2

np.float64(2.048407141795244)

**|t| = 1.090 < t_crit = 2.048 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [11]:
pval2 = 2 * t.sf(abs(t2), df2)  # two-tailed p-value
pval2

np.float64(0.2850015227710625)

**p-value = 0.2847 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [12]:
diff2 = xbar2a - xbar2b           # observed difference in means
lo2 = diff2 - tcrit2 * se2        # lower bound of 95% CI
hi2 = diff2 + tcrit2 * se2        # upper bound of 95% CI
print(f"95% CI for (\u03bc\u2081 - \u03bc\u2082): ({lo2:.4f}, {hi2:.4f})")

95% CI for (μ₁ - μ₂): (-0.3517, 1.1517)


### ✅ Conclusion

| Metric | Value |
|---|---|
| Drug group mean | 5.2 |
| Placebo group mean | 4.8 |
| Pooled SD (sp) | 1.0050 |
| t-statistic | 1.0900 |
| df | 28 |
| Critical value (±) | 2.0484 |
| p-value | 0.2847 |
| 95% CI for difference | (-0.3499, 1.1499) |
| Decision | Fail to reject H₀ |

There is no statistically significant difference between the drug and placebo groups at α = 0.05. The mean difference of 0.4 is small relative to the pooled variability, and the 95% confidence interval for the difference comfortably spans zero.

---

## 📌 Question 3 : Welch's t-Test vs Student's t-Test

Group A: n₁ = 12, x̄₁ = 45.3, s₁ = 8.7  
Group B: n₂ = 30, x̄₂ = 41.8, s₂ = 15.2  
(a) Check equality of variances (Levene's test or variance ratio) and choose the appropriate t-test with justification.  
(b) Compute Welch's degrees of freedom and perform a two-sided Welch's t-test at α = 0.05: state hypotheses, compute test statistic, compute critical t, compute p-value, and state the conclusion.  
(c) Construct a 95% confidence interval for (μ₁ − μ₂).  
(d) State the modern consensus on when to use Welch's t-test.

In [13]:
n3a = 12        # Group A sample size
x3a = 45.3      # Group A sample mean
s3a = 8.7       # Group A sample standard deviation
n3b = 30        # Group B sample size
x3b = 41.8      # Group B sample mean
s3b = 15.2      # Group B sample standard deviation
alpha3 = 0.05   # significance level

#### Part (a) : Variance Check

#### Why this test

Before committing to a two-sample t-test, the variance ratio is inspected. A ratio greater than 2 indicates meaningfully unequal variances, making Welch's t-test the safer choice. The pooled (Student's) t-test assumes equal variances, so it should not be used when that assumption is suspect.

#### Variance Ratio

$$\text{Variance Ratio} = \frac{\max(s_1^2,\, s_2^2)}{\min(s_1^2,\, s_2^2)}$$

In [14]:
var3a = s3a**2                              # Group A variance
var3b = s3b**2                              # Group B variance
ratio3 = max(var3a, var3b) / min(var3a, var3b)  # variance ratio
ratio3

3.0524507861012027

In [15]:
print("Variance A:", var3a)                          # Group A variance
print("Variance B:", var3b)                          # Group B variance
print("Variance Ratio:", ratio3)                     # ratio of larger to smaller

if ratio3 > 2:
    print("Ratio > 2, variances unequal, use Welch's t-test") 
else:
    print("Ratio <= 2, variances similar, Student's t-test possible")  

Variance A: 75.68999999999998
Variance B: 231.04
Variance Ratio: 3.0524507861012027
Ratio > 2, variances unequal, use Welch's t-test


### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| Group A variance | 75.69 |
| Group B variance | 231.04 |
| Variance ratio | 3.05 |
| Test chosen | Welch's t-test |

The variance ratio of 3.05 exceeds 2, indicating that the equal-variance assumption is not tenable. Welch's t-test is used for the remainder of the analysis.

#### Part (b) : Welch's t-Test

#### Step 1 : Hypotheses

- **H₀ :** μ₁ = μ₂ (Group A and Group B have the same population mean)
- **H₁ :** μ₁ ≠ μ₂ (the two group means differ)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

Two independent groups with unknown and unequal population variances call for Welch's t-test. Unlike the pooled t-test, Welch's method does not require equal variances and adjusts the degrees of freedom using the Welch-Satterthwaite approximation, which accounts for the variance imbalance between groups.

#### Step 3 : Test Statistic

$$t = \frac{\bar{x}_1 - \bar{x}_2}{\sqrt{\dfrac{s_1^2}{n_1} + \dfrac{s_2^2}{n_2}}}$$

Welch-Satterthwaite degrees of freedom:

$$df = \frac{\left(\dfrac{s_1^2}{n_1} + \dfrac{s_2^2}{n_2}\right)^2}{\dfrac{\left(\dfrac{s_1^2}{n_1}\right)^2}{n_1 - 1} + \dfrac{\left(\dfrac{s_2^2}{n_2}\right)^2}{n_2 - 1}}$$

In [16]:
se3 = np.sqrt(s3a**2 / n3a + s3b**2 / n3b)  # Welch standard error
t3 = (x3a - x3b) / se3                        # Welch t-statistic
t3

np.float64(0.9351193853114768)

In [17]:
num3 = (var3a / n3a + var3b / n3b)**2                          # numerator of Welch-Satterthwaite formula
den3 = (var3a / n3a)**2 / (n3a - 1) + (var3b / n3b)**2 / (n3b - 1)  # denominator
df3 = num3 / den3                                               # raw degrees of freedom
df3_ceil = math.ceil(df3)                                       # round up per convention
df3_ceil

35

#### Step 4 : Critical Value Method

In [18]:
tcrit3 = t.ppf(1 - alpha3 / 2, df3)  # two-tailed critical value at alpha=0.05
tcrit3

np.float64(2.0308187174932497)

**|t| = 0.935 < t_crit = 2.031 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [19]:
pval3 = 2 * t.sf(abs(t3), df3)  # two-tailed p-value
pval3

np.float64(0.3561964916843625)

**p-value = 0.3562 > 0.05 → Fail to reject H₀.**

### ✅ Part (b) Conclusion

| Metric | Value |
|---|---|
| Group A mean | 45.3 |
| Group B mean | 41.8 |
| t-statistic | 0.9351 |
| df (Welch, ceiling) | 35 |
| Critical value (±) | 2.0308 |
| p-value | 0.3562 |
| Decision | Fail to reject H₀ |

The t-statistic of 0.94 does not exceed the critical value of 2.03, and the p-value of 0.36 is well above α = 0.05. There is no statistically significant difference between the group means.

#### Part (c) : Confidence Interval

#### Step 6 : Confidence Interval

In [20]:
margin3 = tcrit3 * se3             # margin of error
lo3 = (x3a - x3b) - margin3        # lower bound of 95% CI
hi3 = (x3a - x3b) + margin3        # upper bound of 95% CI
print(f"95% CI for (\u03bc\u2081 \u2212 \u03bc\u2082): ({lo3:.4f}, {hi3:.4f})")

95% CI for (μ₁ − μ₂): (-4.1010, 11.1010)


### ✅ Part (c) Conclusion

| Metric | Value |
|---|---|
| Mean difference (x̄₁ - x̄₂) | 3.5 |
| 95% CI for (μ₁ - μ₂) | (-4.1010, 11.1010) |
| Contains zero | Yes |

The 95% confidence interval spans zero, consistent with the hypothesis test result. A true mean difference of zero is plausible given the data.

#### Part (d) : Modern Consensus on Welch's t-Test

The modern consensus, as summarized by Delacre, Lakens, and Leys (2017), is to default to Welch's t-test rather than testing for equal variances first and then choosing. The reasons are:

- When variances are equal, Welch's t-test performs nearly as well as Student's t-test with negligible loss of power.
- When variances are unequal, Student's t-test can produce inflated Type I error rates, especially when sample sizes also differ.
- Pre-testing for equal variances (such as Levene's or Bartlett's) adds an extra layer of uncertainty and does not reliably protect against the error inflation in Student's test.
- Welch's t-test is robust across both equal and unequal variance conditions, making it the safer all-purpose choice.

The practical takeaway: unless there is a domain-specific reason to assume equal variances, Welch's t-test is the appropriate default for comparing two independent group means.

### ✅ Part (d) Conclusion

| Scenario | Recommended test |
|---|---|
| Equal variances assumed | Student's pooled t-test |
| Unequal variances or unsure | Welch's t-test (default) |
| Unequal variances and unequal sample sizes | Welch's t-test (strongly preferred) |

Welch's t-test is the safer default. It costs almost nothing when variances happen to be equal, and it protects Type I error control when they are not.

---

## 📌 Question 4 : Paired t-Test vs Independent t-Test

25 students were tested **before** and **after** a coaching program:

| | Before | After |
|---|---|---|
| Mean (x̄) | 68 | 74 |
| SD (s) | 10 | 12 |
| n | 25 | 25 |

Pearson correlation between before and after scores: **r = 0.72**

**(a)** Compute SE of the mean difference using the paired formula:
$$SE_d = \frac{\sqrt{s_1^2 + s_2^2 - 2 \cdot r \cdot s_1 \cdot s_2}}{\sqrt{n}}$$

**(b)** Run a paired t-test, then repeat as an independent t-test (deliberately ignoring pairing). Compare both.

**(c)** Show quantitatively how much power is lost by ignoring the correlation.

In [ ]:
n4    = 25    # number of pairs (students)
xb4a  = 68    # before-group mean
s4a   = 10    # before-group standard deviation
xb4b  = 74    # after-group mean
s4b   = 12    # after-group standard deviation
r4    = 0.72  # Pearson correlation between paired observations
alpha4 = 0.05 # significance level
d4    = xb4b - xb4a  # mean difference: after minus before
df4   = n4 - 1       # degrees of freedom for paired test
d4

#### Part (a) : SE of the Mean Difference

#### Why this formula

When two measurements come from the same subjects, the variance of their difference is not simply s₁² + s₂². The correlation between the two sets reduces it. The paired SE formula subtracts the shared variance term 2·r·s₁·s₂, so a higher correlation → smaller SE → more powerful test.

#### Formula

$$SE_d = \frac{\sqrt{s_1^2 + s_2^2 - 2 \cdot r \cdot s_1 \cdot s_2}}{\sqrt{n}}$$

In [ ]:
se4p = np.sqrt(s4a**2 + s4b**2 - 2 * r4 * s4a * s4b) / np.sqrt(n4)  # paired SE
se4p

### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| s₁² + s₂² | 244 |
| 2·r·s₁·s₂ | 172.8 |
| Variance of difference | 71.2 |
| Paired SE | 1.1976 |

The correlation of 0.72 removes 172.8 out of 244 units of combined variance, cutting the SE roughly in half compared to what an independent-samples formula would give.

#### Part (b) : Paired t-Test

#### Step 1 : Hypotheses

- **H₀ :** μ_d = 0 (coaching has no effect on mean score)
- **H₁ :** μ_d ≠ 0 (coaching changes mean score)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

The same 25 students are measured twice, so the two sets of scores are not independent. A paired t-test works directly on the within-subject differences, cancelling out individual baseline variation. This is more powerful than an independent t-test whenever the two measurements are positively correlated.

#### Step 3 : Test Statistic

$$t = \frac{\bar{d}}{SE_d} = \frac{\bar{x}_2 - \bar{x}_1}{SE_d}$$

Degrees of freedom: df = n − 1 = 24

In [ ]:
t4p = d4 / se4p  # paired t-statistic: mean difference divided by paired SE
t4p

#### Step 4 : Critical Value Method

In [ ]:
tcrit4p = t.ppf(1 - alpha4 / 2, df4)  # two-tailed critical value, df=24
tcrit4p

**|t| = 5.010 > t_crit = 2.064 → Reject H₀.**

#### Step 5 : p-value Method

In [ ]:
pval4p = 2 * t.sf(abs(t4p), df4)  # two-tailed p-value, df=24
pval4p

**p-value < 0.001 → Reject H₀.**

#### Step 6 : Confidence Interval

In [ ]:
lo4p = d4 - tcrit4p * se4p  # lower bound of 95% CI for mean difference
hi4p = d4 + tcrit4p * se4p  # upper bound of 95% CI for mean difference
print(f"95% CI for mean difference: ({lo4p:.4f}, {hi4p:.4f})")

### ✅ Part (b) Conclusion : Paired t-Test

| Metric | Value |
|---|---|
| Mean difference (d̄) | 6 |
| Paired SE | 1.1976 |
| t-statistic | 5.0096 |
| df | 24 |
| Critical value (±) | 2.0639 |
| p-value | < 0.001 |
| 95% CI for d̄ | (3.5298, 8.4702) |
| Decision | Reject H₀ |

The paired test detects a highly significant improvement. The 95% CI lies entirely above zero, confirming coaching raised scores by somewhere between 3.5 and 8.5 points.

#### Part (b) continued : Independent t-Test (Ignoring Pairing)

#### Step 1 : Hypotheses

Same as above. H₀: μ₁ = μ₂, H₁: μ₁ ≠ μ₂. Two-tailed at α = 0.05.

The only difference is that the pairing is deliberately discarded.

#### Step 2 : Why this is wrong here

Treating paired data as two independent samples ignores the positive correlation between scores. The independent SE formula does not subtract the covariance term, so it overestimates noise. This inflates the SE, shrinks the t-statistic, and can cause a real effect to go undetected.

#### Step 3 : Test Statistic

$$SE_{\text{ind}} = \sqrt{\frac{s_1^2}{n} + \frac{s_2^2}{n}}$$

$$t = \frac{\bar{x}_2 - \bar{x}_1}{SE_{\text{ind}}}$$

Degrees of freedom: df = n₁ + n₂ − 2 = 48

In [ ]:
se4i  = np.sqrt(s4a**2 / n4 + s4b**2 / n4)  # independent SE, no covariance term
df4i  = n4 + n4 - 2                           # degrees of freedom for two-sample test
t4i   = d4 / se4i                             # independent t-statistic
se4i, t4i

#### Step 4 : Critical Value Method

In [ ]:
tcrit4i = t.ppf(1 - alpha4 / 2, df4i)  # two-tailed critical value, df=48
tcrit4i

**|t| = 1.897 < t_crit = 2.011 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [ ]:
pval4i = 2 * t.sf(abs(t4i), df4i)  # two-tailed p-value, df=48
pval4i

**p-value = 0.0635 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [ ]:
lo4i = d4 - tcrit4i * se4i  # lower bound of 95% CI (independent)
hi4i = d4 + tcrit4i * se4i  # upper bound of 95% CI (independent)
print(f"95% CI (independent): ({lo4i:.4f}, {hi4i:.4f})")

### ✅ Part (b) Full Comparison

| Metric | Paired t-test | Independent t-test |
|---|---|---|
| SE | 1.1976 | 3.1623 |
| t-statistic | 5.0096 | 1.8974 |
| df | 24 | 48 |
| Critical value (±) | 2.0639 | 2.0106 |
| p-value | < 0.001 | 0.0635 |
| 95% CI for difference | (3.5298, 8.4702) | (-0.3657, 12.3657) |
| Decision | **Reject H₀** | **Fail to reject H₀** |

The same data, the same effect (d̄ = 6), and opposite conclusions, solely because the independent test discards the correlation structure. The independent CI also spans zero, while the paired CI stays entirely positive.

#### Part (c) : Power Loss from Ignoring Correlation

#### Why this matters

Power is the probability of correctly detecting a true effect. When the SE is inflated by ignoring correlation, the t-statistic shrinks and the test becomes less likely to reject H₀ even when the effect is real. The ratio of the two SEs, and equivalently the two t-statistics, quantifies exactly how much is lost.

#### Formula

The SE ratio follows directly from the two formulas:

$$\frac{SE_{\text{ind}}}{SE_{\text{paired}}} = \frac{\sqrt{\dfrac{s_1^2 + s_2^2}{n}}}{\dfrac{\sqrt{s_1^2 + s_2^2 - 2rs_1s_2}}{\sqrt{n}}} = \sqrt{\frac{s_1^2 + s_2^2}{s_1^2 + s_2^2 - 2rs_1s_2}}$$

The t-statistic ratio is the reciprocal (since both use the same d̄).

In [ ]:
ratio_se4 = se4i / se4p   # SE inflation factor when pairing is ignored
ratio_t4  = t4p  / t4i   # t-statistic deflation factor when pairing is ignored
print(f"SE ratio  (independent / paired):  {ratio_se4:.4f}")
print(f"t-ratio   (paired / independent):  {ratio_t4:.4f}")

### ✅ Part (c) Conclusion

| Metric | Paired | Independent | Ratio |
|---|---|---|---|
| SE | 1.1976 | 3.1623 | ×2.640 (independent is larger) |
| t-statistic | 5.0096 | 1.8974 | ÷2.640 (paired is larger) |
| p-value | < 0.001 | 0.0635 | N/A |

Ignoring the correlation inflates the SE by a factor of 2.64 (a 164% increase) and deflates the t-statistic by the same factor. Analytically, the variance saved by pairing is 2·r·s₁·s₂ = 2 × 0.72 × 10 × 12 = 172.8, which is 70.8% of the total combined variance (244). The remaining 29.2% is what the independent test treats as the only noise. Discarding that 70.8% is what turns a p < 0.001 result into a borderline p = 0.063.